### DISCUSSION TITLE
Parse Markdown in Stack Overflow content

### DISCUSSION THREAD

**Discussion by ctjlewis** [lowlighter/metrics#236] [Ideas]
This library is awesome, and the only use I have for my SO account is to link it here at this point, but the content is not so legible as it frequently contains embedded code:

![image](https://user-images.githubusercontent.com/1657236/115127342-61884580-9f9b-11eb-8089-6650361d1ab3.png)

Not super trivial though: triple caret blocks need to be coerced to inline and trimmed if they run over the limit (as in the second code block in the screenshot), *then* parsed. I would say that second code block could still be included, just with the ellipsis: `interface myInterface { [key: strin...`.

Thoughts?

**Comment by lowlighter** [comment]
Thanks for your suggestion! 

It should be possible 🙂 

Recently [marked.js](https://github.com/markedjs/marked) was added for [metrics insights](https://metrics.lecoq.io/about/) to parse markdown so it could be handled quite easily I guess. As for text formatting, a few CSS rules from the activity plugin can be reused.

Answers are not really "trimmed" actually, it's handled through CSS through with `text-overflow: ellipsis` and a [semi-dirty hack](https://github.com/lowlighter/metrics/blob/f451085c98ac80f521bf8ea20fb62e39046523ae/source/templates/classic/partials/stackoverflow.ejs#L45-L49) 😅. It should already cover the case you mention, though maybe it'll render with "`interface myInterface { [key: strin`..." instead (i.e. ellipsis outside of code block) 

Code hightlighting may be possible too without being too tricky, since [prismjs](https://github.com/prismjs/prism) is already used on the website

As for triple backticks, I'm not sure whether they should be inlined or not 🤔 
It may render well in case of single lines like in your example case, but maybe multi-line answers would be weird?

**Reply by ctjlewis** [reply]
Something to think about is that we can manually specify how wide something should be in terms of characters with the `ch` character unit, even let users override, like so:

```css
/** 80 characters wide */
width: 80ch;
```

Example: Setting `<p>` blocks in this response to `width: 40ch`.

![image](https://user-images.githubusercontent.com/1657236/115129709-0e6bbe00-9fae-11eb-8bf5-9d2c9847f8e2.png)

So basically, if we know how many characters are in each line, and we know how many lines we want, we know how to trim the string length before formatting it, and can work around that first issue.

As for the multi-line code blocks, yeah my impression was they should be inlined in a preview display like that and trimmed. But maybe it would be nicer to start a multiline block and just trim it:

```typescript
interface myInterface {
  [key: strin...
```

Honestly, that does look pretty fancy, and you get the formatting on everything in the code block.

**Reply by lowlighter** [reply]
I've merged #237 which to be able to parse markdown and make code highlighting in it.

I had some trouble with special stackoverflow syntax as code blocks use html comments (`<!-- language: lang-xxx -->`) instead of triple backticks (mainly for runnable snippets).

Currently it'll look like this, which does look nicer.

<img width="400" src="https://user-images.githubusercontent.com/22963968/115147296-a921d680-a05a-11eb-817f-bb7b4a09d4ba.png">

There are still some display issues that needs to be addressed, such as:
- text ellipsis is gone when last line is a code snippet
- ~headers are rendered inline (looks weird)~
- ~css for token highlighting is pretty basic and some colors does not blend well on dark modes~ 
 
___

> Example: Setting `<p>` blocks in this response to width: 40ch.

I didn't know about this unit, it's actually pretty cool!
It would be definitely be more optimized and would reduce SVG filesize but I can't find an easy way to extract `innerText` (to skip html tags), crop text, and "re-html" text as just cropping text would probably leave unclosed tags. 

If it does not require too much work (and does not seems too hacky), we could switch to it instead of using `line-clamp` property

**Reply by ctjlewis** [reply]
Absolutely incredible work @lowlighter, seriously top-notch.

**Reply by lowlighter** [reply]
If you want, you can beta-test this feature by using `@master` branch instead of `@latest` in your workflow. 
Don't hesitate to report any bugs you found if you end up switching to it 🙂

**Reply by ctjlewis** [reply]
[It looks *so* good!](https://raw.githubusercontent.com/ctjlewis/ctjlewis/3f4f60bba1c8fef1dbbd6d11958fbde61178623b/github-metrics.svg)

One thing I did catch is that it HTML-encoded `=>` to `=&gt;` in answer 2 here:

![image](https://user-images.githubusercontent.com/1657236/115478962-eb8b1500-a20c-11eb-8308-ac99c81e8908.png)

**Reply by lowlighter** [reply]
Should be fixed with 1b7ff7e 👍
